### PR TITLE
Safe Spark SQL modules

### PR DESCRIPTION
**Pull request by buonuomo**
This is my rewrite of all the `SQL` modules in sparkle for the safe interface. Currently everything compiles, but I still need to test a few things, so it's not ready to merge yet.

### CODE REVIEW

File: src/linear-types/Control/Distributed/Spark/Safe/SQL/Dataset.hs [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+-- TODO: was it safe to get rid of the unsafePerformIO?
+getEncoder :: forall a. Dataset a %1 -> IO (Encoder a)
+getEncoder (Dataset dsref) = Linear.do
+    UnsafeUnrestrictedReference klass <- findClass $ referenceTypeName (Java.SClass "org.apache.spark.sql.Dataset")
+    Ur fID <- getFieldID klass "org$apache$spark$sql$Dataset$$encoder"
+                     (Java.signature (sing :: Sing ('Iface "org.apache.spark.sql.Encoder")))
+    (ref, obj) <- getObjectField dsref fID
+    deleteLocalRef ref
+    deleteLocalRef klass
+    pure . Encoder . (unsafeCast :: JObject %1 -> J ('Iface "org.apache.spark.sql.Encoder")) $ obj
```
Comment (facundominguez) [inline src/linear-types/Control/Distributed/Spark/Safe/SQL/Dataset.hs:72]:
```Haskell
  getEncoder :: Dataset a %1 -> IO (Encoder a)
￼ getEncoder dsref = Linear.do
    Encoder <$> [java| $dsref.org$apache$spark$sql$Dataset$$encoder |]
```
or if that doesn't work, maybe
```Haskell
   getEncoder :: Dataset a %1 -> IO (Encoder a)
  ￼getEncoder dsref = Linear.do
     Encoder <$> [java| $dsref.getClass().getDeclaredField("org$apache$spark$sql$Dataset$$encoder").get($dsref) |]
```

File: src/linear-types/Control/Distributed/Spark/Safe/SQL/StructType.hs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      (jfields :: J ('Array ('Class "org.apache.spark.sql.types.StructField")))
+    (jfields'', fieldsList) <- foldM 
+                                (\(arr, acc) i -> second ((: acc) . StructField) <$> getObjectArrayElement arr i) 
+                                (jfields', []) 
+                                [0 .. n P.- 1]
+    deleteLocalRef jfields''
+    pure fieldsList
+  where
+    second :: (n %1 -> o) %1 -> (m, n) %1 -> (m, o)
+    second g (a, b) = (a, g b)
```
Comment (facundominguez) [inline src/linear-types/Control/Distributed/Spark/Safe/SQL/StructType.hs:61]: hm, all of the functions here seem to look better with quasiquotations.

**Review by facundominguez — COMMENTED** [review]
Looks good!

You might want to use more quasiquotations at a few places.

This PR is also missing some tests. Maybe converting some of the example programs to use linear types would suffice (e.g. rddops and dataframe).

**Comment by buonuomo** [comment]
Something I realized a week or two ago while converting the dataframe example program to use the safe interface here: sometimes we do have `RDD`'s (or `Dataset`'s) of reference types, most notable an `RDD Row` or `Dataset Row`. In such a case, we're not able to `collect` from such an `RDD`, as there's no `Reify` instance for `Row` (or any safe reference types, I believe). The fact that there's no `Reify` instance is a good thing, as otherwise, we'd be able to get an `Ur Row`, which is not safe, but this does reveal the need for a separate class of collection functions that return linearly bound reference types.

As it stands, there are not too many such functions in `sparkle`, so simply adding analogs like `collectJ` for each one will probably accomplish what we need to for now with minimal conceptual overhead and a fixed amount of boilerplate per relevant function. At first glance, it seems like a type-class solution _might_ be able to eliminate these near-duplicate functions, but I suspect that doing so would adversely affect the comprehensibility of the relevant type signatures in nontrivial ways. 

I'll publish a draft of this solution (along with the example dataframe program using it) soon, in order to demonstrate my proposed fix.

**Comment by facundominguez** [comment]
Sounds good, Noah! Thanks for the update.

**Comment by buonuomo** [comment]
Okay, I've finally added `collectJ` and its companions. In addition I've ported the dataframe example program over to the safe interface, and it appears to work fine. However, I've run into a hurdle in porting over rdd-ops, which is that we don't have any safe reify or reflect instances for streams, meaning that `reduce` doesn't work. At the moment, I'm not sure if writing these instances will simply involve wrapping the unsafe instances, or if it will require porting the entirety of `jvm-streaming` to use linear types.

File: apps/dataframe-safe/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  deps = [
+    "//:sparkle-lib",
+    "@maven//:org_apache_spark_spark_sql_2_11",
+    "@io_tweag_inline_java//jvm",
+	"@stackage//:base",
+	"@stackage//:distributed-closure",
+	"@stackage//:text",
+        "@stackage//:linear-base",
+        "@io_tweag_inline_java//jni",
+        "@io_tweag_inline_java//:inline-java",
```
Comment (facundominguez) [inline apps/dataframe-safe/BUILD.bazel:23]:
```suggestion
    "@io_tweag_inline_java//jvm",
    "@stackage//:base",
    "@stackage//:distributed-closure",
    "@stackage//:text",
    "@stackage//:linear-base",
    "@io_tweag_inline_java//jni",
    "@io_tweag_inline_java//:inline-java",
```

**Comment by facundominguez** [comment]
> At the moment, I'm not sure if writing these instances will simply involve wrapping the unsafe instances, or if it will require porting the entirety of jvm-streaming to use linear types.

I think wrapping the unsafe interfaces could be a fair stop gap. Not wrapping will require porting `jvm-streaming` which is small, but it depends on `jvm-batching` and will be some effort to port the two (not a lot of effort, but still some).

**Pull request merged**